### PR TITLE
Curb instrumentation: tests and CHANGELOG entry

### PR DESCRIPTION
**Pull request by fallwith**
* update `CHANGELOG.md` to announce the new community driven Curb bug fix
* raise Curb `instrumentation.rb` test coverage to 100%/100%

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,7 +9,11 @@
     Thank you to @Arkham for bringing this to our attention!
 
 
+  * **BugFix: Fix memory leak in the Curb instrumentation
 
+    With [PR #1518](https://github.com/newrelic/newrelic-ruby-agent/pull/1518) community member [charkost](https://github.com/charkost) was able to rework the `on_failure` callback logic prepped via the agent's Curb instrumentation in order to avoid some nesting that was causing memory leaks.
+
+    Many thanks, [charkost](https://github.com/charkost)!
```
[the remark this answers is not in the document]
Comment (kaylareopelle) [reply CHANGELOG.md:16]: With all these recommendations, perhaps it's time to add a changelog style guide section to our dev best practices doc.

File: test/multiverse/suites/curb/curb_chain_test.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+require 'curb'
+require 'newrelic_rpm'
+
+class CurbChainTest < Minitest::Test
```
[the remark this answers is not in the document]
Comment (kaylareopelle) [reply test/multiverse/suites/curb/curb_chain_test.rb:8]: Also, how does this work when the tests are run with the `prepend` instrumentation method? Should we have a condition excluding this test when the instrumentation method is `prepend`?

File: test/multiverse/suites/curb/curb_test.rb [hunk 3]
```diff
@@ -145,6 +145,116 @@ def test_method_with_tracing_passes_the_verb_downstream
...
+      wrapped_request = MiniTest::Mock.new
+      wrapped_request.expect :type, nil
+      wrapped_request.expect :uri, nil
+      wrapped_request.expect :method, nil
+      [wrapped_request, nil]
+    end
+    NewRelic::Agent::Tracer.stub :start_external_request_segment, segment do
+      multi.hook_pending_request(request)
+    end
+    request.verify
```
Comment (kaylareopelle) [inline test/multiverse/suites/curb/curb_test.rb:207]: I'm still trying to wrap my head around the `Minitest::Mock#verify` method. Why would we call `#verify` on `request`, but not on `segment`?

**Review by kaylareopelle — DISMISSED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,7 +9,11 @@
     Thank you to @Arkham for bringing this to our attention!
 
 
+  * **BugFix: Fix memory leak in the Curb instrumentation
 
+    With [PR #1518](https://github.com/newrelic/newrelic-ruby-agent/pull/1518) community member [charkost](https://github.com/charkost) was able to rework the `on_failure` callback logic prepped via the agent's Curb instrumentation in order to avoid some nesting that was causing memory leaks.
```
[the remark this answers is not in the document]
Comment (fallwith) [reply CHANGELOG.md]: I've been looking at too many video game titles, like OutRun. Addressed with 99ec7ce94f9d1f34ca1f5295bf99aba14acdd34d

File: CHANGELOG.md [hunk 5]
```diff
@@ -9,7 +9,11 @@
     Thank you to @Arkham for bringing this to our attention!
 
 
+  * **BugFix: Fix memory leak in the Curb instrumentation
```
[the remark this answers is not in the document]
Comment (fallwith) [reply CHANGELOG.md]: I've been looking at too many video game titles, like OutRun. Addressed with 99ec7ce94f9d1f34ca1f5295bf99aba14acdd34d

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fallwith) [reply CHANGELOG.md]: `@` added with 99ec7ce94f9d1f34ca1f5295bf99aba14acdd34d

Good call on discussing / creating a guide

File: test/multiverse/suites/curb/curb_chain_test.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (fallwith) [reply test/multiverse/suites/curb/curb_chain_test.rb:8]: > To my knowledge, we haven't written chain-specific instrumentation tests in the past. Do you think this is something we should add for all of our instrumentation?

I still wonder about the how far the need for chaining extends. Is it critical for 2 instrumentations but utterly disused by all the rest? Was it critical with a much older agent version but used less know? We need to get some test coverage over all areas of the code base, and I'm down for either adding tests or shrinking the surface area in need of testing.

> Also, how does this work when the tests are run with the prepend instrumentation method? Should we have a condition excluding this test when the instrumentation method is prepend?

This is a good idea. I'll see what our options are.

File: test/multiverse/suites/curb/curb_test.rb (hunk 3 above)
Comment (fallwith) [reply test/multiverse/suites/curb/curb_test.rb:207]: The omission of a `#verify` call on `segment` here was just an oversight on my part. Added it with 953b230bece1a2d450474b8e0d77ce9802010dcb.

MiniTest (unlike RSpec and perhaps unlike Mocha) will not automatically verify the mocks at the end of the test method invocation. As a result, `#expect` works like an RSpec `allow` up until `#verify` is used to actually verify that calls were made.

In this PR, my goal is to add `#verify` calls to every mock. The only ones that don't have `#verify` are the really funky ones inside instance level method definitions, and for those which aren't the focus on what's being verified, I'm ok with them working like "allow".

File: test/multiverse/suites/curb/curb_chain_test.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (fallwith) [reply test/multiverse/suites/curb/curb_chain_test.rb]: (We discussed this one synchronously as a team) I have removed the chain tests from this PR with 6d7de4115db52f28443e0d07a3ff8ec8c3de7c08. I had misunderstood what the chain instrumentation method context of multiverse was already giving us. The new tests were a) redundant and b) causing conflicts with prepend, so they have simply been removed.

**Review by tannalynn — APPROVED** [review]
🎉

**Pull request merged**